tikv does not build on ppc64le

1. What version of Rust are you using (`rustc --version`)?
Beta 1.19 and Nightly 1.20 downloaded from - https://static.rust-lang.org/dist/rust-beta-powerpc64le-unknown-linux-gnu.tar.gz

2. What operating system and processor architecture are you using?
ppc64le, RedHat 7.3

3. What did you do?
Build tikv using make / make static_release
The build fails with the following error
error[E0432]: unresolved import `types::os::arch::c95::*`
   --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.1.12/rust/src/liblibc/lib.rs:100:9
    |
100 | pub use types::os::arch::c95::*;
    |         ^^^^^^^^^^^^^^^^^^^^^^^^ Could not find `arch` in `os`
... ...

4. Analysis so far:
Initially, I was suspecting that the libc crate is not ported for ppc64le, however, after interacting with the rust community, I found that support for ppc64le is available in the latest version of the libc crate.
https://github.com/rust-lang/rust/issues/42858

They suggest upgrading utime, nix and other tikv dependencies to the latest version for better platform compatibility. 

I am currently trying to understand what all might need to be updated- any pointers/thoughts on this would be of great help - thanks in advance!!

Regards,
Priya

Hi @siddontang, thanks for your inputs!

I tried this and it fails while compiling **strsim** and **flat_map**

Here is what I did (on a fresh checkout and clean environment)

bash-4.2#  cargo update -p libc
error: There are multiple `libc` packages in your project, and the specification `libc` is ambiguous.
Please re-run this command with `-p <spec>` where `<spec>` is one of the following:
  libc:0.1.12
  libc:0.2.21

bash-4.2# cargo update -p libc:0.1.12
    Updating registry `https://github.com/rust-lang/crates.io-index`
    Updating git repository `https://github.com/pingcap/kvproto.git`
    Updating git repository `https://github.com/pingcap/grpc-rs.git`
    Updating git repository `https://github.com/pingcap/rust-rocksdb.git`
    Updating git repository `https://github.com/pingcap/tipb.git`
    Updating git repository `https://github.com/pingcap/rust-prometheus.git`
    Updating git repository `https://github.com/busyjay/jemallocator.git`

First few lines of the error:
   Compiling strsim v0.6.0
error[E0432]: unresolved import `types::os::arch::c95::*`
   --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.1.12/rust/src/liblibc/lib.rs:100:9
    |
100 | pub use types::os::arch::c95::*;
    |         ^^^^^^^^^^^^^^^^^^^^^^^^ Could not find `arch` in `os`

and last few
error[E0204]: the trait `Copy` may not be implemented for this type
   --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.1.12/rust/src/liblibc/lib.rs:441:26
    |
441 |                 #[derive(Copy, Clone)]
    |                          ^^^^
...
444 |                     pub ifa_name: *mut c_char,
    |                     ------------------------- this field does not implement `Copy`

   Compiling flat_map v0.0.4
error: aborting due to previous error(s)

Rust and cargo versions on my system are:
bash-4.2# rustc --version
rustc 1.20.0-nightly (859c3236e 2017-06-26)
bash-4.2# cargo --version
cargo 0.21.0-nightly (d26fd6f08 2017-06-20)

I will continue to dig into this.

Hi @siddontang,, just a quick update.

I updated strsim, flat_map and a few more dependencies using cargo update. I could proceed, however now I am getting an error while compiling nix and simply updating that using cargo update does not help.

bash-4.2# cargo update -p nix:0.5.1
    Updating registry `https://github.com/rust-lang/crates.io-index`
bash-4.2# cargo update -p nix:0.6.0
    Updating registry `https://github.com/rust-lang/crates.io-index`
bash-4.2# make
cargo build --release --features "default"
   Compiling libc v0.1.12
   Compiling nix v0.5.1
   Compiling librocksdb_sys v0.1.0 (https://github.com/pingcap/rust-rocksdb.git#f7c020d8)
   Compiling grpc-sys v0.1.0 (https://github.com/pingcap/grpc-rs.git#5e425f27)
   Compiling sys-info v0.4.1
   Compiling nix v0.6.0
error[E0432]: unresolved import `types::os::arch::c95::*`
   --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.1.12/rust/src/liblibc/lib.rs:100:9
    |
100 | pub use types::os::arch::c95::*;
    |         ^^^^^^^^^^^^^^^^^^^^^^^^ Could not find `arch` in `os`

error[E0432]: unresolved import `types::os::arch::c99::*`

Looks like the latest version of nix 0.8.1, https://crates.io/crates/nix should work as it depends on libc 0.2.21 - will check how to update nix, next

@siddontang 
Thanks! I think I am hitting another problem on ppc64le ....

Just to be sure, I created a clean setup and tried the above steps i.e.
1.Remove Cargo.lock
2.Use 0.2.24 for libc in Cargo.toml  libc = "0.2.24" 
3.Remove  features = ["static-link"]  in Cargo.toml for  [dependencies.grpc] 

However, it failed at the below point   Compiling ordermap v0.2.10
error[E0432]: unresolved import `types::os::arch::c95::*`
   --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.1.12/rust/src/liblibc/lib.rs:100:9
    |

 I am suspecting that this is because there are other dependencies in Cargo.toml that in turn depend on libc (specifically libc-0.1.12) and will need to be updated as well for the build to work.

That is what I have been trying, but looks like there are multiple such dependencies and it will take some effort to figure them all out, unless there is a way to do a recursive update.

Just a quick update on where I am with the port on ppc64le ...

I was able to resolve mio issues by using the latest version of mio 0.6.3 with the deprecated APIs. Subsequently, tikv compilation was failing due to use of jemalloc, which I have temporarily tried to disable by disabling memory profiling. Now I am getting the following issue 
   Compiling tikv v0.0.1 (file:///root/tikv)
error[E0599]: no associated item named `CeilInt` found for type `tipb::expression::ScalarFuncSig` in the current scope
   --> src/coprocessor/select/xeval/evaluator.rs:555:13
    |
555 |             ScalarFuncSig::CeilInt => self.ceil_int(ctx, expr),
    |             ^^^^^^^^^^^^^^^^^^^^^^

error[E0599]: no associated item named `FloorInt` found for type `tipb::expression::ScalarFuncSig` in the current scope
   --> src/coprocessor/select/xeval/evaluator.rs:557:13
    |
557 |             ScalarFuncSig::FloorInt => self.floor_int(ctx, expr),
    |             ^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

error: Could not compile `tikv`.

To learn more, run the command again with --verbose.

Looking into it further.
Meanwhile the changes for sys-info-rs have been up-streamed and the community has released a new version of the crate.

Thanks @siddontang for the details.

For this particular issue (ppc64le support), apart from the open PR, there is one last issue (jemalloc issue mentioned above) that I am now trying to resolve (had by-passed this call on ppc64le to try and move ahead). Everything, apart from this seems to be working 

  = note: /usr/bin/ld: /root/tikv/target/debug/deps/tikv_server-9c7aaf9d3bc56a45.tikv_server0.rust-cgu.o: undefined reference to symbol 'malloc_stats_print'
          //usr/lib/powerpc64le-linux-gnu/libjemalloc.so.1: error adding symbols: DSO missing from command line
          collect2: error: ld returned 1 exit status


error: aborting due to previous error

error: Could not compile `tikv`.

@BusyJay - quick update - this is looking quite good, so far - I just had to disable the sse flag for rocksdb to get it to build (no other changes were required) for "make build". For, "make test", set AFL_NO_X86=1, first 800+ tests passed before execution was terminated with following error 

test coprocessor::test_select::test_output_offsets ... ok
**E0123 09:06:46.438678648   31242 ev_posix.cc:227]            No event engine could be initialized from epollex
error: process didn't exit successfully: `/tikv/target/debug/deps/integrations-79c58c087a6b4954 --nocapture` (signal: 6, SIGABRT: process abort signal)
Makefile:95: recipe for target 'test' failed
make: *** [test] Error 101**

--- a/components/test_util/src/lib.rs
+++ b/components/test_util/src/lib.rs
@@ -34,19 +34,28 @@ pub use macros::*;
 pub use security::*;
 
 pub fn setup_for_ci() {
-    if env::var("CI").is_ok() && env::var("LOG_FILE").is_ok() {
-        logging::init_log_for_test();
+    if env::var("CI").is_ok() {
+        if env::var("LOG_FILE").is_ok() {
+            logging::init_log_for_test();
+        }
+
+        // HACK! Use `epollex` as the polling engine for gRPC when running CI tests on
+        // Linux and it hasn't been set before.
+        // See more: https://github.com/grpc/grpc/blob/v1.17.2/src/core/lib/iomgr/ev_posix.cc#L124
+        // See more: https://grpc.io/grpc/core/md_doc_core_grpc-polling-engines.html
+        #[cfg(target_os = "linux")]
+        {
+            if env::var("GRPC_POLL_STRATEGY").is_err() {
+                env::set_var("GRPC_POLL_STRATEGY", "epollex");
+            }
+        }
     }
+
     if env::var("PANIC_ABORT").is_ok() {
         // Panics as aborts, it's helpful for debugging,
         // but also stops tests immediately.
         tikv::util::set_panic_hook(true, "./");
     }
-
-    // HACK! Always use epollex in tests on Linux.
-    // See more: https://github.com/grpc/grpc/blob/v1.17.2/src/core/lib/iomgr/ev_posix.cc#L124
-    #[cfg(target_os = "linux")]
-    env::set_var("GRPC_POLL_STRATEGY", "epollex");
 
     tikv::util::check_environment_variables();
 }

